File: Programacion_II/Python/ahorcado/ahorcado.py
import sys
from random import *


def menu() -> int:
    op = 0
    while op != 1 and op != 2 and op != 3:
        clear()
        if op != 0:
            print("Ingrese una opción correcta")
        print("Bienvenido al ahorcado")
        op = input(
            "1. Jugar contra la máquina\n2. Jugar contra otro jugador\n3. Salir\nIngrese una opción: ")
        if op.isdigit():
            op = int(op)
        else:
            op = -1
    return op


def clear():
    print("\n"*90)


def lineas_archivo(archivo) -> int:
    i = 0
    for l in archivo:
        i += 1
    return i


def elegir_palabra(archivo, cant_lineas: int) -> str:
    archivo.seek(0)

    j = 1
    num_pal = randint(1, cant_lineas)

    while j != num_pal:
        archivo.readline()
        j += 1

    palabra_elegida = archivo.readline()
    palabra_elegida = palabra_elegida[:len(palabra_elegida) - 1]

    return palabra_elegida


def entrada(elegidas: set) -> str:
    letra = '1'
    while not letra.isalpha():
        letra = input("Ingrese una letra: ")
        if letra in elegidas:
            print("Letra ya elegida")
            letra = '1'
        if len(letra) > 1:
            print("Ingrese solo un caracter")
            letra = '1'
        else:
            print("Ingrese un caracter válido")
    return letra


def letras_correctas(palabra: str) -> dict[str, list[int]]:
    dic = {}
    i = 0
    for l in palabra:
        if l not in dic.keys():
            dic[l] = [i]
        else:
            dic[l].append(i)
        i += 1
    return dic


def actualizar_tablero(l: str, letras: dict, adivinanza) -> list[str]:
    for i in letras[l]:
        adivinanza[i] = l
    return adivinanza


def mostrar_tablero(adivinanza: list[str], vidas: int, elegidas: set):
    for e in adivinanza:
        print(e, end=' ')
    print("Vidas:", vidas)
    print("Letras elegidas:", end=' ')
    for e in elegidas:
        print(e, end=' ')
    print()


def jugar(palabra_secreta: str) -> tuple[bool, str]:
    correctas = letras_correctas(palabra_secreta)
    cant_diferentes = len(correctas.keys())
    vidas = 6
    adivinanza = ["_"] * len(palabra_secreta)
    letras_elegidas = set()

    while vidas > 0 and cant_diferentes != 0:
        mostrar_tablero(adivinanza, vidas, letras_elegidas)
        letra = entrada(letras_elegidas)
        letras_elegidas.add(letra)
        if letra in correctas.keys():
            adivinanza = actualizar_tablero(letra, correctas, adivinanza)
            cant_diferentes -= 1
        else:
            vidas -= 1
        clear()

    if vidas > 0:
        return (True, palabra_secreta)
    return (False, palabra_secreta)


def ingresar_palabra() -> str:
    incorrecto = 1
    palabra = ""
    while incorrecto:
        incorrecto = 0
        palabra = input("Ingrese la palabra secreta:")
        for letra in palabra:
            if not letra.isalpha():
                incorrecto = 1
    return palabra


def mensaje_resultado(resultado: tuple[bool, str]):
    print("Ganaste, la palabra era" if resultado[0] else "Perdiste, la palabra era", f'"{
          resultado[1]}"')
    input("Presione enter para continuar:")


def main(lemario):
    modo = menu()
    palabra_secreta = ''
    while modo != 3:
        if modo == 1:
            archivo = open(lemario, "r")
            largo = lineas_archivo(archivo)
            clear()
            palabra_secreta = elegir_palabra(archivo, largo)

        if modo == 2:
            palabra_secreta = ingresar_palabra()

        resultado = jugar(palabra_secreta)
        mensaje_resultado(resultado)
        modo = menu()


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Ingresar el archivo lemario.")
        exit()
    main(sys.argv[1])

File: Programacion_II/C/ahorcado/ahorcado.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <time.h>
#include <ctype.h>

#define LARGO_MAXIMO 50

int numero_random(int num){
    srand(time(0));
    return random() % num;
}

void clear(){
    for(int i = 99; i; i--)
        printf("\n");
}

int cant_lineas(FILE* archivo){
    fseek(archivo, 0, 0);
    char linea[LARGO_MAXIMO];
    int cant_lineas = 0;
    while(fscanf(archivo, "%s\n", linea) != EOF)
        cant_lineas++; 

    return cant_lineas;
}

char* elegir_palabra(FILE* archivo, int cant_lineas){
    fseek(archivo, 0, 0);
    int linea_random = numero_random(cant_lineas);
    char linea[LARGO_MAXIMO];
    while(linea_random-- && (fscanf(archivo, "%s\n", linea) != EOF));
    char* palabra_elegida = malloc(sizeof(char) * (strlen(linea) + 1));
    strcpy(palabra_elegida, linea);

    return palabra_elegida;
}

void mostrar_palabra(char* palabra){
    for(int i = 0; palabra[i] != '\0'; i++)
        printf("%c ", palabra[i]);
}

char entrada(int* intentos){
    int c = '0';
    while(c < 97 || c > 122 || intentos[c % 26]){
        printf("Ingrese una letra: "); 
        c = tolower(scanf("%d", &c));
        if(c < 97 || c > 122 )
            printf("Ingrese un caracter valido.\n");
        if(intentos[c % 26])
            printf("Ingrese una letra que no haya ingresado");
        //clear();
    }

    return c;
}

int comprobar_letra(char letra, char* adivinanza, char* secreto){
    int acierto = 0;
    for(int i = 0; i < strlen(secreto); i++){
        if(letra == secreto[i]){
            adivinanza[i] = letra;
            acierto = 1;
        }
    }
    
    return acierto;
} 


int vs_maquina(char* secreto){
    int i = 0, largo = strlen(secreto), vidas = 7, intentos[26] = {};
    char* adivinanza = malloc(sizeof(char) * (largo+1));
    
    for(i = 0; i < largo; i++)
        adivinanza[i] = '_';
    adivinanza[i] = '\0';

    while(strcmp(adivinanza, secreto) != 0 && vidas){
        mostrar_palabra(adivinanza);
        char letra = entrada(intentos);
        printf("letra: %c", letra);
        intentos[letra % 26] = 1;
        if(!comprobar_letra(letra, adivinanza, secreto))
            vidas--;
    }

    return (vidas > 0);
}

int main(int argc, char const *argv[]){
    if(argc != 2){
        printf("Ingrese el archivo lemario.\n");
        return 1;
    }

    FILE* archivo = fopen(argv[1], "r");
    assert(archivo != NULL);
    int cant_lineas_archivo = cant_lineas(archivo);
    char* palabra_secreta = elegir_palabra(archivo, cant_lineas_archivo); 
    vs_maquina(palabra_secreta);

    free(palabra_secreta);


    return 0;
}
